src/Parsers/ASTFunction.cpp

```diff
@@ -300,66 +347,123 @@ namespace
+                markSecretArgument(5);
             }
         }
 
-        std::pair<size_t, size_t> findS3TableEngineSecretArguments() const
+        void findS3TableEngineSecretArguments()
         {
+            if (isCollectionNameArgument(0))
+            {
+                /// S3(named_collection, ..., secret_access_key = 'secret_access_key')
+                findSecretNamedArgument("secret_access_key", 1);
+                return;
+            }
+
             /// We replace 'aws_secret_access_key' with '[HIDDEN'] for the following signatures:
             /// S3('url', 'aws_access_key_id', 'aws_secret_access_key', 'format')
             /// S3('url', 'aws_access_key_id', 'aws_secret_access_key', 'format', 'compression')
 
             /// But we should check the number of arguments first because we don't need to do that replacements in case of
             /// S3('url' [, 'format' [, 'compression']])
             if (arguments->size() < 4)
-                return npos;
+                return;
 
-            return {2, 3};
+            markSecretArgument(2);
         }
 
-        std::pair<size_t, size_t> findDatabaseEngineSecretArguments() const
+        void findDatabaseEngineSecretArguments()
         {
             const String & engine_name = function.name;
             if ((engine_name == "MySQL") || (engine_name == "MaterializeMySQL") ||
                 (engine_name == "MaterializedMySQL") || (engine_name == "PostgreSQL") ||
                 (engine_name == "MaterializedPostgreSQL"))
             {
                 /// MySQL('host:port', 'database', 'user', 'password')
-                /// PostgreSQL('host:port', 'database', 'user', 'password', ...)
-                return {3, 4};
+                /// PostgreSQL('host:port', 'database', 'user', 'password')
+                findMySQLDatabaseSecretArguments();
+            }
+        }
+
+        void findMySQLDatabaseSecretArguments()
+        {
+            if (isCollectionNameArgument(0))
+            {
+                /// MySQL(named_collection, ..., password = 'password', ...)
+                findSecretNamedArgument("password", 1);
             }
             else
             {
-                return npos;
+                /// MySQL('host:port', 'database', 'user', 'password')
+                markSecretArgument(3);
             }
         }
 
-        std::pair<size_t, size_t> findBackupNameSecretArguments() const
+        void findBackupNameSecretArguments()
         {
             const String & engine_name = function.name;
             if (engine_name == "S3")
             {
                 /// BACKUP ... TO S3(url, [aws_access_key_id, aws_secret_access_key])
-                return {2, 3};
+                markSecretArgument(2);
             }
-            else
+        }
+
+        /// Whether a specified argument can be the name of a named collection?
+        bool isCollectionNameArgument(size_t arg_idx) const
```

Comment:
maybe you can call it `isNamedCollectionArgument` ? as in most places its referred to as `named_collection`. What do you think?

Reply:
In such cases all the arguments are related to the named collection:

```
func(named_collection_name, named_collection_override_param1 = val1, named_collection_override_param2 = val2, ...)
```

I think I'll rename the function to `isNamedCollectionNameArgument()` - it seems a bit long, but correct.